esx_service/vmdk_ops.py

```diff
@@ -1052,25 +1052,36 @@ def set_vol_opts(name, options):
        msg = 'Volume {0} not found.'.format(vol_name)
        logging.warning(msg)
        return False
-       
+   
     # For now only allow resetting the access mode.
-    valid_opts = [kv.ACCESS]
+    valid_opts = [kv.ACCESS, kv.ATTACH_AS]
     invalid = frozenset(opts.keys()).difference(valid_opts)
     if len(invalid) != 0:
         msg = 'Invalid options: {0} \n'.format(list(invalid)) \
                + 'Options that can be edited: ' \
                + '{0}'.format(list(valid_opts))
         raise ValidationError(msg)
-
-    if not opts[kv.ACCESS] in kv.ACCESS_TYPES:
-       msg = 'Invalid option value {0}.\n'.format(opts[kv.ACCESS]) +\
-             'Supported values are {0}.\n'.format(kv.ACCESS_TYPES)
-       logging.warning(msg)
-       return False
+    if kv.ACCESS in opts.keys():    
+        if not opts[kv.ACCESS] in kv.ACCESS_TYPES:
+            msg = 'Invalid option value {0}.\n'.format(opts[kv.ACCESS]) +\
+                    'Supported values are {0}.\n'.format(kv.ACCESS_TYPES)
+            logging.warning(msg)
+            return False
+
+    if kv.ATTACH_AS in opts.keys():    
+        if not opts[kv.ATTACH_AS] in kv.ATTACH_AS_TYPES:
+            msg = 'Invalid option value {0}.\n'.format(opts[kv.ATTACH_AS]) +\
+                    'Supported values are {0}.\n'.format(kv.ATTACH_AS_TYPES)
+            logging.warning(msg)
+            return False
+   
```

Comment:
This code  block isn't easy to extend if each option needs to repeat the above steps. Instead,

a. change valid opts to a dict of {'<opt>': [valid choices], ....} then

b. loop over the keys in the provided options and see if its present in valid_opts.heys{}. If yes, then
     c. check if the option value is in the list of valid choices for that option. 

     d. If not then
            e. prepare an error message and remember that the option value is invalid.

            f. Exit the loop

g. Figure if there was an error in parsing the options and option values. If yes then log and exit.

Else the above code blocks will get duplicated over and over for each new optioin thats added to the set command.